src/Update/Nix/FetchGit.hs

```diff
@@ -1,122 +1,66 @@
 updatesFromFile :: FilePath -> IO (Either Warning [SpanUpdate])
 updatesFromFile filename = do
   t <- T.readFile filename
   case parseNixTextLoc t of
-    Failure d -> pure $ Left (CouldNotParseInput d)
-    Success e -> case fetchGitUpdateInfos t e of
-      Left w -> pure (Left w)
-      Right updateInfos ->
-        sequenceA2 <$> mapConcurrently2 updateInfoToUpdate updateInfos >>= \case
-          Left w -> pure $ Left w
-          Right pairs -> pure $ Right (universeBi pairs)
-
--- | Given an expression, return all the opportunities for updates for any
--- subexpression.
-fetchGitUpdateInfos :: Text
-                    -> NExprLoc
-                    -> Either Warning [UpdateSource FetchGitUpdateInfo]
-fetchGitUpdateInfos t e = do
-  -- Get the 'FetchGitArgs' values from the tree
-  argTrees <- exprToUpdateSource e
-  -- Get the 'FetchGitUpdateInfo' values
-  traverse2 (fetchGitArgsToUpdate t) argTrees
+    Failure parseError -> pure $ Left (CouldNotParseInput parseError)
+    Success expr -> case exprToFetchTree expr of
+      Left scanError -> pure (Left scanError)
+      Right treeWithArgs ->
+        sequenceA <$> mapConcurrently getFetchGitLatestInfo treeWithArgs >>= \case
+          Left getLatestInfoError -> pure $ Left getLatestInfoError
+          Right treeWithLatest -> pure $ return (fetchTreeToSpanUpdates treeWithLatest)
 
 --------------------------------------------------------------------------------
--- Extracting places to update from the AST
+-- Extracting information about fetches from the AST
 --------------------------------------------------------------------------------
 
--- | A fetchgit, fetchgitPrivate or fetchFromGitHub value's expressions
-data FetchGitArgs = FetchGitArgs{ repoLocation :: RepoLocation
-                                , revExpr      :: NExprLoc
-                                , sha256Expr   :: NExprLoc
-                                }
-  deriving (Show, Data)
-
--- Get a set of 'UpdateSource' trees from a nix expression.
-exprToUpdateSource :: NExprLoc -> Either Warning [UpdateSource FetchGitArgs]
-exprToUpdateSource = sequenceA . para exprChildrenToUpdateSources
-
--- Given an expression and the update sources contained in its children, return
--- a new set of update sources
-exprChildrenToUpdateSources :: NExprLoc
-                            -> [[Either Warning (UpdateSource FetchGitArgs)]]
-                            -> [Either Warning (UpdateSource FetchGitArgs)]
-exprChildrenToUpdateSources e subss =
-  let subs = concat subss
-  in case e of
-
-      -- If this is a call to some symbol
+-- Get a FetchTree from a nix expression.
+exprToFetchTree :: NExprLoc -> Either Warning (FetchTree FetchGitArgs)
+exprToFetchTree = para exprToFetchTreeCore
+
+exprToFetchTreeCore :: NExprLoc
+                    -> [Either Warning (FetchTree FetchGitArgs)]
+                    -> Either Warning (FetchTree FetchGitArgs)
+exprToFetchTreeCore e subs =
+  case e of
```

Comment:
It would be good to keep the indentation style consistent. Sadly there's no good tool for this in Haskell.

The code indents patterns two spaces past the beginning of the `case` keyword at the moment.